File: t3_content_generation/t2/gpt_image_generation.py
import base64
from datetime import datetime

from commons.constants import OPENAI_HOST
from t3_content_generation._openai_client import OpenAIClientT3


# https://platform.openai.com/docs/guides/image-generation?image-generation-model=gpt-image-1&api=image&multi-turn=imageid
# ---
# Request:
# curl -X POST "https://api.openai.com/v1/images/generations" \
#     -H "Authorization: Bearer $OPENAI_API_KEY" \
#     -H "Content-type: application/json" \
#     -d '{
#         "model": "gpt-image-1",
#         "prompt": "smiling catdog."
#     }'
# Response:
# {
#   "created": 1699900000,
#   "data": [
#     {
#       "b64_json": Qt0n6ArYAEABGOhEoYgVAJFdt8jM79uW2DO...,
#     }
#   ]
# }


def main(model_name: str, request: str, **kwargs):
    client = OpenAIClientT3(
        endpoint=OPENAI_HOST+"/v1/images/generations",
    )

    response = client.call(
        model=model_name,
        prompt=request,
        **kwargs
    )
    image_base64 = response["data"][0]["b64_json"]

    image_bytes = base64.b64decode(image_base64)
    filename= f"{datetime.now()}.png"
    with open(filename, "wb") as f:
        f.write(image_bytes)

    print(f"Image saved as {filename}")


main(
    model_name="gpt-image-1",
    request="smiling catdog"
)

File: tasks/src/t3/content/generation/t2/GptImageGeneration.java
package t3.content.generation.t2;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import commons.Constants;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Base64;

/**
 * T3-2b: GPT-Image-1 Image Generation
 * <p>
 * Generates an image via /v1/images/generations using gpt-image-1.
 * Unlike DALL-E 3, the response returns the image as base64 JSON (b64_json)
 * rather than a URL — this implementation decodes and saves it as a PNG file.
 */
public class GptImageGeneration {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final HttpClient HTTP = HttpClient.newHttpClient();

    public static void main(String[] args) throws Exception {
        // JSON template for the request body
        String body = """
                {
                    "model": "gpt-image-1",
                    "prompt": "Smiling catdog"
                }
                """;

        System.out.println("REQUEST:\n" + body);

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(Constants.OPENAI_IMAGES_GENERATIONS_ENDPOINT))
                .header("Authorization", "Bearer " + Constants.OPENAI_API_KEY)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(body))
                .build();

        HttpResponse<String> response = HTTP.send(request, HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() != 200) {
            throw new RuntimeException("HTTP " + response.statusCode() + ": " + response.body());
        }

        JsonNode responseJson = MAPPER.readTree(response.body());
        System.out.println("RESPONSE:\n" + MAPPER.writerWithDefaultPrettyPrinter().writeValueAsString(responseJson));

        // Decode the base64 image data
        String imageBase64 = responseJson.path("data").path(0).path("b64_json").asText();
        byte[] imageBytes = Base64.getDecoder().decode(imageBase64);

        // Determine the output directory (t3_content_generation/t2/)
        Path outputDir = Path.of("tasks/src/t3/content/generation/t2/");

        String filename = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd_HH-mm-ss")) + ".png";
        Path outputPath = outputDir.resolve(filename);
        
        Files.write(outputPath, imageBytes);
        System.out.println("Image saved as " + outputPath.toAbsolutePath());
    }
}
